# 2. Data Loading & Quality Checks

In [2]:
# FIRST let's import the packages we need:
import pandas as pd # for data manipulation
import numpy as np # for linear algebra and conditionals
import seaborn as sns # for statistical visuals
import scipy.stats # for regression

In [3]:
from google.colab import files
uploaded = files.upload()

Saving Campaign-Data.csv to Campaign-Data.csv


In [4]:
data = pd.read_csv('Campaign-Data.csv')
data.columns

Index(['Client ID', 'Client Type', 'Number of Customers', 'Montly Target',
       'Zip Code', 'Calendardate', 'Amount Collected', 'Unit Sold',
       'Campaign (Email)', 'Campaign (Flyer)', 'Campaign (Phone)',
       'Sales Contact 1', 'Sales Contact 2', 'Sales Contact 3',
       'Sales Contact 4', 'Sales Contact 5', 'Number of Competition'],
      dtype='object')

In [5]:
data.head()

Client ID      Client Type  Number of Customers  Montly Target  Zip Code  \
0  ID-987275  Medium Facility               2800.0          125.0    1003.0   
1  ID-987275  Medium Facility               2800.0          125.0    1003.0   
2  ID-987275  Medium Facility               2800.0          125.0    1003.0   
3  ID-987275  Medium Facility               2800.0          125.0    1003.0   
4  ID-987275  Medium Facility               2800.0          125.0    1003.0   

  Calendardate  Amount Collected  Unit Sold  Campaign (Email)  \
0   16-01-2014               0.0        0.0               0.0   
1   16-02-2014         3409460.0       24.0               0.0   
2   18-03-2014        10228384.0       75.0               0.0   
3   18-04-2014        17047304.0      123.0               0.0   
4   19-05-2014        23866224.0      171.0               0.0   

   Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  Sales Contact 2  \
0               0.0               0.0              0.0              0.0   
1               0.0               0.0              0.0              0.0   
2               0.0               0.0              0.0              0.0   
3               0.0               0.0              0.0        3547500.0   
4               0.0               0.0              0.0              0.0   

   Sales Contact 3  Sales Contact 4  Sales Contact 5 Number of Competition  
0              0.0              0.0              0.0                   Low  
1              0.0              0.0         322500.0                   Low  
2              0.0              0.0              0.0                   Low  
3        1290000.0              0.0              0.0                   Low  
4              0.0              0.0              0.0                   Low

In [6]:
# Amount Collected represents the target variable that we are going to use to help define our marketing strategy
# We will use the variables: Campaign (Email) through Sales Contact 5 as our independent variables to help formulate the marekting strategy per client type

# 3. Feature Additions & Engineering

In [7]:
# We need to ensure our calendar column has the correct date/make sure any "String" is changed into a "Date"
data['Calendardate']=pd.to_datetime(data['Calendardate'])
# We are then going to use dt to create additional columns (one for the month and one for the year)
data['Calendar_Month']=data['Calendardate'].dt.month
data['Calendar_Year']=data['Calendardate'].dt.year

<ipython-input-7-b0e159427bcd>:2: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data['Calendardate']=pd.to_datetime(data['Calendardate'])


# 4. Exploratory Data Analysis & Statistical Analysis

<b> 4.1 Exploring and Understanding Basic Data </b>

1. Distribution of Data across different accounts
2. Difference of Sales in Account Types (Using Categorical Mean)

<b> 4.2 Statistical Analysis - Answering the Questions</b>
1. Impact of Marketing Strategy on Sales (Using Correlation, Regression and Decision Tree)
2. Impact of Competition on Sales
3. How different types of client can have different strategies (Broken down Question 1 and Question 2 based on Account Type)

# 4.1 Exploring and Understanding Basic Data

In [8]:
data['Client Type'].value_counts(normalize=True)

Client Type
Large Facility      0.459677
Small Facility      0.282258
Medium Facility     0.169355
Private Facility    0.088710
Name: proportion, dtype: float64

In [9]:
# This function in pandas gives the relative frequency (proportions) of each category in the Client Type column

**What this tells us:**


*   **Large Facility (45.97%):** About 46% of the clients are large facilities.

*   **Small Facility (28.23%):** Around 28% of the clients are small facilities.

*   **Medium Facility (16.94%):** About 17% of the clients are medium-sized facilities.

*   **Private Facility (8.87%):** Only about 9% of the clients are private facilities.

  This distribution gives an understanding of the composition of the dataset in terms of client types, with the largest group being large facilities, followed by small and medium facilities, and finally, a smaller percentage of private facilities.







In [10]:
pd.crosstab(data['Number of Competition'],data['Client Type'],margins=True,normalize='columns')

Client Type            Large Facility   Medium Facility  Private Facility  \
Number of Competition                                                       
High                          0.166667         0.166667          0.166667   
Low                           0.833333         0.833333          0.833333   

Client Type            Small Facility        All  
Number of Competition                             
High                          0.166667  0.166667  
Low                           0.833333  0.833333

In [11]:
# The pd.crosstab function with the normalize='columns' argument shows the relative frequency of each Number of Competition category within each Client Type column, normalized by the columns

**What this tells us:**


*   **High Competition (0.166667):** For each client type (Large Facility, Medium Facility, Private Facility, Small Facility), 16.67% of them are in the "High" competition category.

*   **Low Competition (0.833333):** For each client type, 83.33% of them are in the "Low" competition category.

  This indicates that, regardless of the client type, most clients are in the "Low" competition category, and a smaller proportion is in the "High" competition category.

  It seems the competition level is heavily skewed toward the "Low" category across all client types.







In [12]:
# Now let's find the mean for each when we group by Number of Competition
data.groupby('Number of Competition').mean('Amount Collected') # I just put 'Amount Collected' in because without it I was getting an error

Number of Customers  Montly Target  Zip Code  \
Number of Competition                                                 
High                           1456.935484      75.080645    1003.0   
Low                            1456.935484      75.080645    1003.0   

                       Amount Collected   Unit Sold  Campaign (Email)  \
Number of Competition                                                   
High                       2.974789e+07  213.127016     105398.938508   
Low                        1.445570e+07  103.132258     150862.165766   

                       Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  \
Number of Competition                                                        
High                      994046.717540      45198.036895    146945.564516   
Low                       623692.979839      26693.304194    128219.758065   

                       Sales Contact 2  Sales Contact 3  Sales Contact 4  \
Number of Competition                                                      
High                      2.685333e+06     1.786754e+06     72172.379032   
Low                       1.890916e+06     1.883634e+06     70481.854839   

                       Sales Contact 5  Calendar_Month  Calendar_Year  
Number of Competition                                                  
High                       8452.620968            10.5         2015.0  
Low                       15864.919355             5.7         2014.4

**What this tells us:**

*   for 'Amount Collected', the average is higher for High competition, despite there being more clients in the Low competition.
*   Same for Units Sold



In [13]:
# Now let's find the mean for each when we group by Client Type
data.groupby('Client Type').mean('Amount Collected') # I just put 'Amount Collected' in because without it I was getting an error

Number of Customers  Montly Target  Zip Code  \
Client Type                                                      
Large Facility            1380.842105      71.578947    1003.0   
Medium Facility           3940.761905     202.857143    1003.0   
Private Facility           400.727273      20.454545    1003.0   
Small Facility             422.514286      21.285714    1003.0   

                  Amount Collected   Unit Sold  Campaign (Email)  \
Client Type                                                        
Large Facility        1.999880e+07  143.098684     142273.609649   
Medium Facility       4.075997e+07  290.583333     437217.097817   
Private Facility      5.030246e+06   35.784091       5183.715152   
Small Facility        1.637759e+06   11.689286      11975.986310   

                  Campaign (Flyer)  Campaign (Phone)  Sales Contact 1  \
Client Type                                                             
Large Facility        8.192056e+05      45595.436623    133667.763158   
Medium Facility       1.552603e+06      49176.847619    398645.833333   
Private Facility      2.272919e+05       5522.470455      1221.590909   
Small Facility        9.120875e+04          0.000000      8062.500000   

                  Sales Contact 2  Sales Contact 3  Sales Contact 4  \
Client Type                                                           
Large Facility       2.034013e+06     2.017039e+06    119287.280702   
Medium Facility      4.822783e+06     4.698646e+06     85104.166667   
Private Facility     6.376705e+05     4.434375e+05      3664.772727   
Small Facility       7.617143e+05     3.727946e+05      4223.214286   

                  Sales Contact 5  Calendar_Month  Calendar_Year  
Client Type                                                       
Large Facility       16266.447368             6.5         2014.5  
Medium Facility      33273.809524             6.5         2014.5  
Private Facility     12215.909091             6.5         2014.5  
Small Facility        1535.714286             6.5         2014.5

**What this tells us:**

*   Even though the Medium Facility is only 17% of our client base (ranking 3rd out of all the accounts), it is bringing in the most money and we would want to explore that more to find out *why*.
*  This prompts the question: "What is the marketing strategy for the Medium Facility?"


# 4.2 Statistical Analysis - Answering the Questions

In [14]:
# Now let's dive into our correlation analysis
# First i am going to get only the numeric data
numeric_data = data.select_dtypes(include='number')
# Next i am going to compute correlation
correlation_matrix = numeric_data.corr()
# Now i am goint to inspect the specific correlations for Amount Collected
print(correlation_matrix[['Amount Collected']])

                     Amount Collected
Number of Customers          0.607496
Montly Target                0.608204
Zip Code                          NaN
Amount Collected             1.000000
Unit Sold                    0.997515
Campaign (Email)             0.248235
Campaign (Flyer)             0.444337
Campaign (Phone)             0.034858
Sales Contact 1              0.277478
Sales Contact 2              0.552112
Sales Contact 3              0.357887
Sales Contact 4              0.236165
Sales Contact 5              0.095795
Calendar_Month               0.139425
Calendar_Year                0.286194


**What this tells us:**

*   Remember that correlation goes from -1 to +1

In [15]:
## Consolidated Strategy for Targeting

# Create a light green color palette for the heatmap styling
cm = sns.light_palette("green", as_cmap=True)

# Perform correlation analysis on selected columns
correlation_analysis = pd.DataFrame(data[['Amount Collected',
    'Campaign (Email)', 'Campaign (Flyer)', 'Campaign (Phone)',
    'Sales Contact 1', 'Sales Contact 2', 'Sales Contact 3',
    'Sales Contact 4', 'Sales Contact 5']].corr()['Amount Collected']).reset_index()

# Rename columns for clarity
correlation_analysis.columns = ['Impacting Variable', 'Degree of Linear Impact (Correlation)']

# Remove the 'Amount Collected' row itself from the correlation analysis
correlation_analysis = correlation_analysis[correlation_analysis['Impacting Variable'] != 'Amount Collected']

# Sort the correlation values in descending order
correlation_analysis = correlation_analysis.sort_values('Degree of Linear Impact (Correlation)', ascending=False)

# Apply background gradient for better visualization
correlation_analysis.style.background_gradient(cmap=cm)

**What this tells us:**

*   Sales Contact 2 is highly correlated with Amount Collected (in relation to the other variables)

**1. Impact of Marketing Strategy on Sales (Using Correlation, Regression and Decision Tree)**

In [16]:
# Now let's use the same code structure, but group by Client Type

# Create a light green color palette for the heatmap styling
cm = sns.light_palette("green", as_cmap=True)

# Perform correlation analysis grouped by 'Client Type'
correlation_analysis = pd.DataFrame(data.groupby('Client Type')[['Amount Collected',
    'Campaign (Email)', 'Campaign (Flyer)', 'Campaign (Phone)',
    'Sales Contact 1', 'Sales Contact 2', 'Sales Contact 3',
    'Sales Contact 4', 'Sales Contact 5']].corr()['Amount Collected']).reset_index()

# Sort by 'Client Type' and correlation with 'Amount Collected' in descending order
correlation_analysis = correlation_analysis.sort_values(['Client Type', 'Amount Collected'], ascending=False)

# Rename the columns for better clarity
correlation_analysis.columns = ['Acc Type', 'Variable Impact on Sales', 'Impact']

# Remove the row where 'Variable Impact on Sales' is 'Amount Collected' and reset the index
correlation_analysis = correlation_analysis[correlation_analysis['Variable Impact on Sales'] != 'Amount Collected'].reset_index(drop=True)

# Apply background gradient for better visualization
correlation_analysis.style.background_gradient(cmap=cm)

**What this tells us:**

*   Email campaigns are not as highly correlated to Private and Large Facilities, so those accounts might need to spend less effort on those

In [17]:
# Now let's start to look at more causation and move away from correlations

**Regression Analysis (Market Sales and Strategies)**

In [18]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

# Clean up column names by replacing spaces with underscores and removing parentheses
data.columns = [mystring.replace(" ", "_") for mystring in data.columns]
data.columns = [mystring.replace("(", "") for mystring in data.columns]
data.columns = [mystring.replace(")", "") for mystring in data.columns]

# Fit an OLS regression model using specified variables (y=mx + b, which in our case is: y=mx + mx1 + mx2...+ b)
results = smf.ols('Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
       Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5', data=data).fit()

# Print the regression summary
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:       Amount_Collected   R-squared:                       0.480
Model:                            OLS   Adj. R-squared:                  0.478
Method:                 Least Squares   F-statistic:                     342.1
Date:                Sun, 05 Jan 2025   Prob (F-statistic):               0.00
Time:                        04:29:45   Log-Likelihood:                -54512.
No. Observations:                2976   AIC:                         1.090e+05
Df Residuals:                    2967   BIC:                         1.091e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept        1.481e+06   5.12e+05     

In [19]:
# If the P-value is above .05, the result is not statistically significant and the null hypothesis is not rejected
# So we are just going to filter those out
# First let's format it as an easily-readable table
from io import StringIO

# Extract the HTML content for the regression results table
html_content = results.summary().tables[1].as_html()

# Wrap the HTML content in a StringIO object
html_file = StringIO(html_content)

# Now manually parse the HTML content using pandas' read_html function
df = pd.read_html(html_file, header=0, index_col=0)[0].reset_index()

# Now let's filter
df=df.reset_index()
df=df[df['P>|t|']<0.05][['index','coef']]
df

index          coef
0        Intercept  1.481000e+06
2   Campaign_Flyer  3.337600e+00
4  Sales_Contact_1  4.236800e+00
5  Sales_Contact_2  3.638200e+00
6  Sales_Contact_3  2.343200e+00
7  Sales_Contact_4  1.094780e+01

**Regression Analysis (Market Sales and Strategies) - Broken for different account types**

In [20]:
consolidated_summary = pd.DataFrame()  # Initialize an empty DataFrame to store results

# Loop over each unique account type in the 'Client_Type' column
for acctype in list(set(list(data['Client_Type']))):

    # Filter the data to include only the rows for the current account type
    temp_data = data[data['Client_Type'] == acctype].copy()

    # Fit an OLS regression model for the current account type
    results = smf.ols('Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
    Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5', data=temp_data).fit()

    from io import StringIO

    # Extract the HTML content for the regression results table
    html_content = results.summary().tables[1].as_html()

    # Wrap the HTML content in a StringIO object
    html_file = StringIO(html_content)

    # Now manually parse the HTML content using pandas' read_html function
    df = pd.read_html(html_file, header=0, index_col=0)[0].reset_index()

    # Filter the results for significant variables (p-value < 0.05) and select only relevant columns
    df = df[df['P>|t|'] < 0.05][['index', 'coef']]

    # Rename columns for clarity
    df.columns = ['Variable', 'Coefficent (Impact)']

    # Add a column for the account type to the DataFrame
    df['Account Type'] = acctype

    # Sort the DataFrame by coefficient value in descending order
    df = df.sort_values('Coefficent (Impact)', ascending=False)

    # Remove the intercept from the results
    df = df[df['Variable'] != 'Intercept']

    # Print the current account type for progress tracking
    print(acctype)

    # Print the results for the current account type
    print(df)

    # Append the current results to the consolidated summary using pd.concat
    consolidated_summary = pd.concat([consolidated_summary, df], ignore_index=True)

Small Facility 
          Variable  Coefficent (Impact)     Account Type
5  Sales_Contact_2             0.810100  Small Facility 
3   Campaign_Phone            -0.000003  Small Facility 
Private Facility
          Variable  Coefficent (Impact)      Account Type
5  Sales_Contact_2               6.6223  Private Facility
Medium Facility
          Variable  Coefficent (Impact)     Account Type
2   Campaign_Flyer               4.1059  Medium Facility
5  Sales_Contact_2               3.5778  Medium Facility
4  Sales_Contact_1               3.1365  Medium Facility
6  Sales_Contact_3               2.1174  Medium Facility
Large Facility 
          Variable  Coefficent (Impact)     Account Type
4  Sales_Contact_1              11.6731  Large Facility 
7  Sales_Contact_4              10.6145  Large Facility 
5  Sales_Contact_2               4.0031  Large Facility 
2   Campaign_Flyer               2.7204  Large Facility 
6  Sales_Contact_3               2.0316  Large Facility 
3   Campaign_Phone   

In [21]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
from io import StringIO
import pandas as pd

consolidated_summary = pd.DataFrame()

for acctype in list(set(list(data['Client_Type']))):
    print(acctype)

    # Filter data for the current account type
    temp_data = data[data['Client_Type'] == acctype].copy()

    # Fit OLS regression model
    results = smf.ols('Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
    Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5', data=temp_data).fit()

    # Extract HTML content from regression results
    html_content = results.summary().tables[1].as_html()

    # Wrap HTML content in StringIO
    html_file = StringIO(html_content)

    # Read HTML content into DataFrame
    df = pd.read_html(html_file, header=0, index_col=0)[0].reset_index()

    # Filter significant variables (p-value < 0.05)
    df = df[df['P>|t|'] < 0.05][['index', 'coef']]

    # Rename columns for clarity
    df.columns = ['Variable', 'Coefficent (Impact)']

    # Add Account Type column
    df['Account Type'] = acctype

    # Sort by Coefficient in descending order within the current account type
    df = df.sort_values('Coefficent (Impact)', ascending=False)

    # Remove the intercept
    df = df[df['Variable'] != 'Intercept']

    # Concatenate to the consolidated summary DataFrame
    consolidated_summary = pd.concat([consolidated_summary, df], ignore_index=True)

    # Print the regression results summary for the current account type
    print(results.summary())

# Sort the entire consolidated_summary by Coefficient in descending order,
# but within each Account Type group
consolidated_summary['Account Type'] = consolidated_summary['Account Type'].astype(str)
consolidated_summary = consolidated_summary.sort_values(by=['Account Type', 'Coefficent (Impact)'], ascending=[True, False])

# Print the final sorted consolidated summary
print(consolidated_summary)

Small Facility 
                            OLS Regression Results                            
Dep. Variable:       Amount_Collected   R-squared:                       0.054
Model:                            OLS   Adj. R-squared:                  0.046
Method:                 Least Squares   F-statistic:                     6.829
Date:                Sun, 05 Jan 2025   Prob (F-statistic):           6.63e-08
Time:                        04:29:45   Log-Likelihood:                -14125.
No. Observations:                 840   AIC:                         2.827e+04
Df Residuals:                     832   BIC:                         2.830e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept        8.789e+05

# 5. Final Recommendations

Using the below table we can use the coefficent to see how much return we can derive from each dollar we spend, here we can clearly see that for different account type different Campaigns and Different Sales Contact are effective with different extend.

<b>Case Explanation - Medium Facility </b><br>
For Example Medium Facility shows decent results with Flyer Campiagns and each dollar spend return 4 dollars on average. Sales Contact 2 is highly effective followed by Sales Contact 1 and Sales Contact 3. Else all other strategy shows no impact can be dropped to save cost.

In [22]:
consolidated_summary

Variable  Coefficent (Impact)      Account Type
7   Sales_Contact_1            11.673100   Large Facility 
8   Sales_Contact_4            10.614500   Large Facility 
9   Sales_Contact_2             4.003100   Large Facility 
10   Campaign_Flyer             2.720400   Large Facility 
11  Sales_Contact_3             2.031600   Large Facility 
12   Campaign_Phone            -3.536100   Large Facility 
3    Campaign_Flyer             4.105900   Medium Facility
4   Sales_Contact_2             3.577800   Medium Facility
5   Sales_Contact_1             3.136500   Medium Facility
6   Sales_Contact_3             2.117400   Medium Facility
2   Sales_Contact_2             6.622300  Private Facility
0   Sales_Contact_2             0.810100   Small Facility 
1    Campaign_Phone            -0.000003   Small Facility

In [23]:
# Reset the index of the consolidated summary DataFrame
consolidated_summary.reset_index(inplace=True)

# Drop the old index column
consolidated_summary.drop('index', inplace=True, axis=1)

# Rename the columns for clarity
consolidated_summary.columns = ['Variable', 'Return on Investment', 'Account Type']

# Round the "Return on Investment" values to one decimal place
consolidated_summary['Return on Investment'] = consolidated_summary['Return on Investment'].apply(lambda x: round(x, 1))

# Apply a background gradient to the DataFrame for visualization (e.g., in a Jupyter notebook)
consolidated_summary.style.background_gradient(cmap='RdYlGn')

In [38]:
import matplotlib.pyplot as plt
from io import StringIO
import pandas as pd
import statsmodels.formula.api as smf

# Initialize an empty DataFrame
consolidated_summary = pd.DataFrame()

# Loop through account types and process regression results
for acctype in list(set(list(data['Client_Type']))):
    temp_data = data[data['Client_Type'] == acctype].copy()
    results = smf.ols('Amount_Collected ~ Campaign_Email + Campaign_Flyer + Campaign_Phone +\
    Sales_Contact_1 + Sales_Contact_2 + Sales_Contact_3 + Sales_Contact_4 + Sales_Contact_5',
                      data=temp_data).fit()

    # Extract regression table and parse it
    html_content = results.summary().tables[1].as_html()
    html_file = StringIO(html_content)
    df = pd.read_html(html_file, header=0, index_col=0)[0].reset_index()

    # Filter significant results
    df = df[df['P>|t|'] < 0.05][['index', 'coef']]
    df.columns = ['Variable', 'Return on Investment']
    df['Account Type'] = acctype

    # Append to consolidated summary
    consolidated_summary = pd.concat([consolidated_summary, df], ignore_index=True)

# Remove intercept rows
consolidated_summary = consolidated_summary[consolidated_summary['Variable'] != 'Intercept']

# Convert 'Return on Investment' column to numeric
consolidated_summary['Return on Investment'] = pd.to_numeric(
    consolidated_summary['Return on Investment'], errors='coerce'
)

# Apply monetary formatting
def format(x):
    return "${:.1f}".format(x)

consolidated_summary['Return on Investment'] = consolidated_summary['Return on Investment'].apply(format)

# Rename columns for clarity
consolidated_summary.columns = ['Variable', 'Return on Investment', 'Account Type']

# Apply background gradient
consolidated_summary.style.background_gradient(cmap='RdYlGn')

In [39]:
consolidated_summary.to_csv('consolidated_summary.csv')
files.download('consolidated_summary.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>